drivers/usb/device/usb_dc_numaker.c

```diff
@@ -0,0 +1,1943 @@
+
+	uint8_t usbd_hw_ep_hndl; /* BSP USBD driver EP handle EP0, EP1, EP2, etc */
+	uint32_t usbd_hw_ep_cfg; /* BSP USBD driver EP configuration */
+
+	/* EP DMA buffer */
+	bool dmabuf_valid;
+	uint32_t dmabuf_base;
+	uint32_t dmabuf_size;
+
+	/* On USBD, no H/W FIFO. Simulate based on above DMA buffer with
+	 * one-shot implementation
+	 */
+	uint32_t read_fifo_pos;
+	uint32_t read_fifo_used;
+	uint32_t write_fifo_pos;
+	uint32_t write_fifo_free;
+
+	/* NOTE: On USBD, Setup and CTRL OUT are not completely separated. CTRL OUT MXPLD
+	 * can be overridden to 8 by next Setup. To overcome it, we make one copy of CTRL
+	 * OUT MXPLD immediately on its interrupt.
+	 */
+	uint32_t mxpld_ctrlout;
+
+	/* EP address */
+	bool ep_addr_valid;
+	uint8_t ep_addr;
+
+	/* EP MPS */
+	bool ep_mps_valid;
+	uint16_t ep_mps;
+
+	usb_dc_ep_callback cb; /* EP callback function */
+};
+
+/* EP context manager */
+struct numaker_usbd_ep_mgmt {
+	/* EP context management
+	 *
+	 * Allocate-only, and de-allocate all on re-initialize in usb_dc_attach().
+	 */
+	uint8_t ep_idx;
+
+	/* DMA buffer management
+	 *
+	 * Allocate-only, and de-allocate all on re-initialize in usb_dc_attach().
+	 */
+	uint32_t dmabuf_pos;
+
+	/* Pass Setup packet from ISR to thread */
+	bool new_setup;
+	struct usb_setup_packet setup_packet;
+
+	struct numaker_usbd_ep ep_pool[NUMAKER_USBD_EP_MAXNUM];
+};
+
+/* Mutable device context */
+struct usb_dc_numaker_data {
+	uint8_t addr; /* Host assigned USB device address */
+
+	struct k_mutex sync_mutex;
+
+	/* Enable interrupt top/bottom halves processing
+	 *
+	 * Registered callbacks may use mutex or other kernel functions which are not supported
+	 * in interrupt context
+	 */
+	struct k_msgq msgq;
+	struct numaker_usbd_msg msgq_buf[CONFIG_USB_DC_NUMAKER_MSG_QUEUE_SIZE];
+
+	K_KERNEL_STACK_MEMBER(msg_hdlr_thread_stack,
+			      CONFIG_USB_DC_NUMAKER_MSG_HANDLER_THREAD_STACK_SIZE);
+	struct k_thread msg_hdlr_thread;
+
+	usb_dc_status_callback status_cb; /* Status callback function */
+
+	struct numaker_usbd_ep_mgmt ep_mgmt; /* EP management */
+};
+
+/* Device context */
+struct usb_dc_numaker_device {
+	const struct usb_dc_numaker_config *config;
+	struct usb_dc_numaker_data *data;
+};
+
+/* Function declaration */
+static inline struct usb_dc_numaker_device *usb_dc_numaker_device_get(void);
+
+/* Internal functions */
+
+static inline void numaker_usbd_lock(const struct usb_dc_numaker_device *dev)
+{
+	struct usb_dc_numaker_data *data = dev->data;
+
+	k_mutex_lock(&data->sync_mutex, K_FOREVER);
+}
+
+static inline void numaker_usbd_unlock(const struct usb_dc_numaker_device *dev)
+{
+	struct usb_dc_numaker_data *data = dev->data;
+
+	k_mutex_unlock(&data->sync_mutex);
+}
+
+/* S/W connect */
+static inline void numaker_usbd_sw_connect(struct usb_dc_numaker_device *dev)
+{
+	ARG_UNUSED(dev);
+
+	/* Clear all interrupts first for clean */
+	USBD_CLR_INT_FLAG(USBD_GET_INT_FLAG());
```

Comment:
We do not need another ugly driver.
Please do not mix low level code like `config->usbd_base->ATTR |= USBD_ATTR_RWAKEUP_Msk;` with the obscuring code from your HAL.

The way you write this driver it should rather be:

```c
	const struct usb_dc_numaker_config *config = dev->config;

	config->base->INTSTS = config->base->INTSTS;

	config->base->INTEN = USBD_INT_BUS | USBD_INT_USB |USBD_INT_FLDET |
			      USBD_INT_WAKEUP | USBD_INT_SOF);;
```

Reply:
usually, we reuse hal driver as much as possible, then make direct access to h/w registers when the driver doesn't provide needed functions or doesn't meet use cases.

Reply:
Usually, we prefer native drivers, and we do not mix low level and HAL, and we do not like ugly drivers, and drivers should supports multiple instances (requirement for the new USB device API).

With a little effort, this driver can be made nice, native and prepared for porting to the new API. I show you the way, my NACK stays until you fix it.

Reply:
made this driver native by porting needed functions (mainly macro functions) from HAL, with instance enabled. only macro constants defined in HAL are used.